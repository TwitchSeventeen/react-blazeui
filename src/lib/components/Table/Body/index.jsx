import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Body(props) {
  const {
    children, className, dataName, as: Component, ...other
  } = props;

  const classes = classNames(
    'c-table__body',
    className,
  );

  return (
    <Component
      className={classes}
      {...other}
      data-name={dataName}
    >
      { children }
    </Component>
  );
}

Body.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Body.defaultProps = {
  className: null,
  children: null,
  dataName: 'table-body',
  as: 'tbody',
};

export default Body;
