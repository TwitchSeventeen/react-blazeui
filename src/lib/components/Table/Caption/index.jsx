import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Caption(props) {
  const {
    children, as: Component, className, dataName, ...other
  } = props;

  const classes = classNames(
    'c-table__caption',
    className,
  );

  return (
    <Component
      className={classes}
      {...other}
      data-name={dataName}
    >
      { children }
    </Component>
  );
}

Caption.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Caption.defaultProps = {
  className: null,
  children: null,
  dataName: 'table-caption',
  as: 'caption',
};

export default Caption;
