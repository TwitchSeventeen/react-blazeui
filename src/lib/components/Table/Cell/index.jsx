import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Cell(props) {
  const {
    children, className, dataName, ...other
  } = props;

  const classes = classNames(
    'c-table__cell',
    className,
  );

  return (
    <td
      className={classes}
      {...other}
      data-name={dataName}
    >
      { children }
    </td>
  );
}

Cell.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
};

Cell.defaultProps = {
  className: null,
  children: null,
  dataName: 'table-cell',
};

export default Cell;
