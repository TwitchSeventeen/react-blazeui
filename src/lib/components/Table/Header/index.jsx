import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Header(props) {
  const {
    children, className, ...other
  } = props;

  const classes = classNames(
    'c-table__head',
    className,
  );

  return (
    <thead className={classes} {...other}>
      { children }
    </thead>
  );
}

Header.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
};

Header.defaultProps = {
  className: null,
  children: null,
};

export default Header;
