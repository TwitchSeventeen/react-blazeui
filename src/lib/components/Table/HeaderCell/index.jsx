import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function HeaderCell(props) {
  const {
    children, className, ...other
  } = props;

  const classes = classNames(
    'c-table__cell',
    className,
  );

  return (
    <th className={classes} {...other}>
      { children }
    </th>
  );
}

HeaderCell.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
};

HeaderCell.defaultProps = {
  className: null,
  children: null,
};

export default HeaderCell;
