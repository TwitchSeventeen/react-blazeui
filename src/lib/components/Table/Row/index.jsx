import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Row(props) {
  const {
    children, as: Component, className, head, disabled, clickable, dataName, ...other
  } = props;

  const classes = classNames(
    'c-table__row',
    { 'c-table__row--heading': head },
    { 'c-table__row--disabled': disabled },
    { 'c-table__row--clickable': clickable },
    className,
  );

  return (
    <Component
      className={classes}
      {...other}
      data-name={dataName}
    >
      { children }
    </Component>
  );
}

Row.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Defines this row is a heading
   */
  head: PropTypes.bool,
  /**
   * Defines a disabled row
   */
  disabled: PropTypes.bool,
  /**
   * Defines a clickable row
   */
  clickable: PropTypes.bool,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Row.defaultProps = {
  className: null,
  children: null,
  head: false,
  disabled: false,
  clickable: false,
  dataName: 'table-row',
  as: 'tr',
};

export default Row;
