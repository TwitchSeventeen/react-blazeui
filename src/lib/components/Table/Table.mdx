---
name: Table
route: /components/table
menu: Components
---

import { Playground, PropsTable } from 'docz'
import Table from './index';

# Table

Tables, great for displaying tabular data.

## Basic Table
<Playground>
    <Table>
        <Table.Caption>Basic table</Table.Caption>
        <Table.Header>
            <Table.Row head>
                <Table.HeaderCell>First name</Table.HeaderCell>
                <Table.HeaderCell>Last name</Table.HeaderCell>
                <Table.HeaderCell>Username</Table.HeaderCell>
            </Table.Row>
        </Table.Header>
        <Table.Body>
            <Table.Row>
                <Table.Cell>Jim</Table.Cell>
                <Table.Cell>Kirk</Table.Cell>
                <Table.Cell>@captaink</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Mr.</Table.Cell>
                <Table.Cell>Spock</Table.Cell>
                <Table.Cell>@science101</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Nyota</Table.Cell>
                <Table.Cell>Uhura</Table.Cell>
                <Table.Cell>@comms</Table.Cell>
            </Table.Row>
        </Table.Body>
    </Table>
</Playground>

## Striped Table
<Playground>
    <Table striped>
        <Table.Caption>Striped rows</Table.Caption>
        <Table.Header>
            <Table.Row head>
                <Table.HeaderCell>First name</Table.HeaderCell>
                <Table.HeaderCell>Last name</Table.HeaderCell>
                <Table.HeaderCell>Username</Table.HeaderCell>
            </Table.Row>
        </Table.Header>
        <Table.Body>
            <Table.Row>
                <Table.Cell>Jim</Table.Cell>
                <Table.Cell>Kirk</Table.Cell>
                <Table.Cell>@captaink</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Mr.</Table.Cell>
                <Table.Cell>Spock</Table.Cell>
                <Table.Cell>@science101</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Nyota</Table.Cell>
                <Table.Cell>Uhura</Table.Cell>
                <Table.Cell>@comms</Table.Cell>
            </Table.Row>
        </Table.Body>
    </Table>
</Playground>

## Clickable Table
<Playground>
    <Table clickable>
        <Table.Caption>Clickable table with disabled row</Table.Caption>
        <Table.Header>
            <Table.Row head>
                <Table.HeaderCell>First name</Table.HeaderCell>
                <Table.HeaderCell>Last name</Table.HeaderCell>
                <Table.HeaderCell>Username</Table.HeaderCell>
            </Table.Row>
        </Table.Header>
        <Table.Body>
            <Table.Row disabled>
                <Table.Cell>Jim</Table.Cell>
                <Table.Cell>Kirk</Table.Cell>
                <Table.Cell>@captaink</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Mr.</Table.Cell>
                <Table.Cell>Spock</Table.Cell>
                <Table.Cell>@science101</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Nyota</Table.Cell>
                <Table.Cell>Uhura</Table.Cell>
                <Table.Cell>@comms</Table.Cell>
            </Table.Row>
        </Table.Body>
    </Table>
</Playground>

## Clickable Row
<Playground>
    <Table>
        <Table.Caption>Clickable row, disabled row</Table.Caption>
        <Table.Header>
            <Table.Row head>
                <Table.HeaderCell>First name</Table.HeaderCell>
                <Table.HeaderCell>Last name</Table.HeaderCell>
                <Table.HeaderCell>Username</Table.HeaderCell>
            </Table.Row>
        </Table.Header>
        <Table.Body>
            <Table.Row clickable>
                <Table.Cell>Jim</Table.Cell>
                <Table.Cell>Kirk</Table.Cell>
                <Table.Cell>@captaink</Table.Cell>
            </Table.Row>
            <Table.Row disabled>
                <Table.Cell>Mr.</Table.Cell>
                <Table.Cell>Spock</Table.Cell>
                <Table.Cell>@science101</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Nyota</Table.Cell>
                <Table.Cell>Uhura</Table.Cell>
                <Table.Cell>@comms</Table.Cell>
            </Table.Row>
        </Table.Body>
    </Table>
</Playground>

## Condensed
<Playground>
    <Table condensed>
        <Table.Caption>Small table and text</Table.Caption>
        <Table.Header>
            <Table.Row head>
                <Table.HeaderCell>First name</Table.HeaderCell>
                <Table.HeaderCell>Last name</Table.HeaderCell>
                <Table.HeaderCell>Username</Table.HeaderCell>
            </Table.Row>
        </Table.Header>
        <Table.Body>
            <Table.Row>
                <Table.Cell>Jim</Table.Cell>
                <Table.Cell>Kirk</Table.Cell>
                <Table.Cell>@captaink</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Mr.</Table.Cell>
                <Table.Cell>Spock</Table.Cell>
                <Table.Cell>@science101</Table.Cell>
            </Table.Row>
            <Table.Row>
                <Table.Cell>Nyota</Table.Cell>
                <Table.Cell>Uhura</Table.Cell>
                <Table.Cell>@comms</Table.Cell>
            </Table.Row>
        </Table.Body>
    </Table>
</Playground>

## Properties
### Table
<PropsTable of={Table} />

### Table.Body
<PropsTable of={Table.Body} />

### Table.Caption
<PropsTable of={Table.Caption} />

### Table.Cell
<PropsTable of={Table.Cell} />

### Table.Header
<PropsTable of={Table.Header} />

### Table.HeaderCell
<PropsTable of={Table.HeaderCell} />

### Table.Row
<PropsTable of={Table.Row} />