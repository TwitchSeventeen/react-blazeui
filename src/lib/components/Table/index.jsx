import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Caption from './Caption';
import Header from './Header';
import Row from './Row';
import Cell from './Cell';
import HeaderCell from './HeaderCell';
import Body from './Body';

function Table(props) {
  const {
    children, as: Component, className, striped, clickable, condensed, dataName, ...other
  } = props;

  const classes = classNames(
    'c-table',
    { 'c-table--striped': striped },
    { 'c-table--clickable': clickable },
    { 'c-table--condensed': condensed },
    className,
  );

  return (
    <Component
      className={classes}
      {...other}
      data-name={dataName}
    >
      { children }
    </Component>
  );
}

Table.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Defines a table with striped rows
   */
  striped: PropTypes.bool,
  /**
   * Defines a clickable table
   */
  clickable: PropTypes.bool,
  /**
   * Defines a condensed table
   */
  condensed: PropTypes.bool,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Table.defaultProps = {
  className: null,
  children: null,
  striped: false,
  clickable: false,
  condensed: false,
  dataName: 'table',
  as: 'table',
};

Table.Caption = Caption;
Table.Header = Header;
Table.Row = Row;
Table.Cell = Cell;
Table.HeaderCell = HeaderCell;
Table.Body = Body;

export default Table;
