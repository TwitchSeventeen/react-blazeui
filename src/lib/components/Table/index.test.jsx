import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Table from './index';

describe('Table', () => {
  afterEach(cleanup);

  it('should be defined', () => {
    expect(Table).toBeDefined();
  });

  it('renders correctly', () => {
    const { asFragment } = render(
      <Table>
        lorem ipsum dolor
      </Table>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a condensed table', () => {
    const { asFragment } = render(
      <Table condensed>
        <Table.Caption>Small table and text</Table.Caption>
        <Table.Header>
          <Table.Row head>
            <Table.HeaderCell>First name</Table.HeaderCell>
            <Table.HeaderCell>Last name</Table.HeaderCell>
            <Table.HeaderCell>Username</Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          <Table.Row>
            <Table.Cell>Jim</Table.Cell>
            <Table.Cell>Kirk</Table.Cell>
            <Table.Cell>@captaink</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Mr.</Table.Cell>
            <Table.Cell>Spock</Table.Cell>
            <Table.Cell>@science101</Table.Cell>
          </Table.Row>
          <Table.Row>
            <Table.Cell>Nyota</Table.Cell>
            <Table.Cell>Uhura</Table.Cell>
            <Table.Cell>@comms</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
