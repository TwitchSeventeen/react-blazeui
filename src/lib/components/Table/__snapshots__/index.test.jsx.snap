// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders a condensed table 1`] = `
<DocumentFragment>
  <table
    class="c-table c-table--condensed"
    data-name="table"
  >
    <caption
      class="c-table__caption"
      data-name="table-caption"
    >
      Small table and text
    </caption>
    <thead
      class="c-table__head"
      data-name="table-header"
    >
      <tr
        class="c-table__row c-table__row--heading"
        data-name="table-row"
      >
        <th
          class="c-table__cell"
          data-name="table-headercell"
        >
          First name
        </th>
        <th
          class="c-table__cell"
          data-name="table-headercell"
        >
          Last name
        </th>
        <th
          class="c-table__cell"
          data-name="table-headercell"
        >
          Username
        </th>
      </tr>
    </thead>
    <tbody
      class="c-table__body"
      data-name="table-body"
    >
      <tr
        class="c-table__row"
        data-name="table-row"
      >
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Jim
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Kirk
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          @captaink
        </td>
      </tr>
      <tr
        class="c-table__row"
        data-name="table-row"
      >
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Mr.
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Spock
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          @science101
        </td>
      </tr>
      <tr
        class="c-table__row"
        data-name="table-row"
      >
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Nyota
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          Uhura
        </td>
        <td
          class="c-table__cell"
          data-name="table-cell"
        >
          @comms
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table renders correctly 1`] = `
<DocumentFragment>
  <table
    class="c-table"
    data-name="table"
  />
  lorem
   
  ipsum
   
  dolor
</DocumentFragment>
`;
